proposals/0001-dep-process.md

```diff
@@ -0,0 +1,147 @@
+
+* **Standard** for technical changes to the protocol, on-disk formats, or
+  public APIs.
+* **Process** for formalizing community processes or other (technical or
+  non-technical) decisions. For example, a security vulnerability reporting
+  policy, a process for handling conflicts of interest, or procedures for
+  mentoring new developers.
+
+The status of an DEP can be:
+
+* **Draft**: writen up, open PR for discussion
+* **Active**: PR accepted; adopted or intended for implementation in mainline
+  libraries and clients as appropriate
+* **Closed**: PR was closed without merging; either consensus was against, a
+  decision was postponed, or the authors withdrew their proposal.
+* **Superseded**: a formerly "active" DEP has been made obsolete by a new
+  active DEP; the new DEP should specify specific old DEPs that it would
+  supersede.
+
+A changelog should be kept in the DEP itself giving the date of any changes of
+status.
+
+A template file is provided, but sections can be added or removed as
+appropriate for a specific DEP.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+There are already multiple sources of technical documentation: the Dat
+[protocol website][proto-website], the Dat [whitepaper][whitepaper], Dat
+website [documentation section][docs], the [discussion repo][discussion-repo]
+issues, and the [datprotocol github group][datproto-group] (containing, eg, the
+`dat.json` repo/spec). Without consensus and consolidation, this would be "yet
+another" place to look.
+
+[proto-website]: https://www.datprotocol.com/
+[whitepaper]: https://github.com/datproject/docs/blob/master/papers/dat-paper.md
+[docs]: https://docs.datproject.org/
+[datproto-group]: https://github.com/datprotocol
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+**TODO:**
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Who are "core developers"? What is the specific decision making process for
+accepting or rejecting a given DEP? Optimistically, it would be clear from
+reading a PR discussion thread whether "consensus" has been reached or not, but
+this might be ambiguous or intimidating to first-time contributors.
+
+When are or are not accepted DEPs binding? Presumably a technical DEP should indicate
+
+Are "standard"/technical and "process" the right categories and terms to use?
+Some DEP systems (like the IETF one) also use terminology like "informational"
+(seems like we could rely on blog posts) an "normative". This might be helpful
+for documenting things like peer discovery methods, which have intentially not
+been specified as part of the formal spec (to make it clear that there is no
+single "right way" to do discovery), but they should be documented somewhere.
+
+Some communities eventually subsume the core protocol specification itself into
+the DEP process (for example, the official Bittorrent protocol was documented
+in a "BEP"). Should dat do this? Or rely on the current (and possibly future)
+Dat whitepaper?
```

Comment:
I vote that we put the protocol documentation into the DEP process.

Reply:
Thinking outloud:
I'm not sure whether we should do this as one big document, or split it into components (basically along the node module abstraction layers: sleep, hypercore, hyperdb, discovery). I think the later... and then either have a "how it all fits together" DEP (maybe a "normative" one showing one way it could be done?), or rely on the whitepaper, or hand wave that it's self obvious.
I guess we don't actually need to show how a client would work all together in the specification, and should leave that as undefined as possible?

Reply:
> I guess we don't actually need to show how a client would work all together in the specification, and should leave that as undefined as possible?

Seems like maybe this would fit in better at docs.datproject.org. It can be more hand waving so its easier to keep updated. Otherwise we may have to update with each spec change.

I think the key is that implementation details need to always be 100% up to date. More general, this is how Dat works don't need describe all the details.

Reply:
I like the proposal of having DEPs for each abstraction layer and then an Informative/Descriptive on how they integrate. I think it does need to be a DEP though, because it's pretty important information for somebody making a full implementation. What I'd expect is that the "Fit-together DEP" would give an overview of the different pieces and describe how they interact, which would make it a very useful highlevel description of the protocol. Isolating the subDEPs would help them be specced more completely, and avoid enormous and unwieldly DEP docs.